Fix CA2012 warnings

solyutor opened the pull request:
Another bunch of threading issues fixes. 
A couple of changes should be reviewed carefully:
* `SystemTextJsonSerializer` uses `ArrayPool<byte>.Shared` to buffer data from streams hence it's a  performance penalty for sync versions of 'FromJson()' methods. 
* ~~Inverted sync/async calls for `AsyncLiveAggregatorBase` is a breaking change, so any inheritor will have to override and implement sync `Build()` method~~.  I've figured out the purpose of `AsyncLiveAggregatorBase`. It was absolutely wrong idea. There is no good and safe way to call async code from sync hence I replaced sync method with `throw new NoSupportedException()`. 

I suppose it still deserve merging as it's improves reliability.

oskardudycz commented on src/Marten/Util/SharedBuffer.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace Marten.Util
+{
+    internal readonly struct SharedBuffer : IDisposable
```
**FIX:** @solyutor, looking at the failing tests, you're missing skipping SOH control character (https://github.com/JasperFx/marten/blob/b5325056d6226159565036b0eda89d0b89ff5429/src/Marten/Util/StreamExtensions.cs#L15). 
See [FromJsonAsync](https://github.com/JasperFx/marten/pull/1967/files#diff-860d4464563fb983fc220a5da7bcf182e25876de07eccf88e720e2fb1c578517R101) implementation.

solyutor replied:
Thanks for pointing to the problem. I also missed `_deserializeOptions` and fixed this as well.

oskardudycz commented:
@solyutor thanks for the follow-up PR! See my comment above.

solyutor commented:
I tested it locally on my pg10 and it worked. I'll upgrade my postgres to version 12 and retest it.

oskardudycz commented:
@solyutor, I think that the issue is not with the PG version, but with the serialiser (`System.Text.Json`). To run tests for it, you have to set the `DEFAULT_SERIALIZER` environment variable to `SystemTextJson`. (read more about how this works in my blog post https://event-driven.io/en/how_to_setup_a_test_matrix_in_xunit/).

solyutor commented:
@oskardudycz - thanks, it helped a lot.  I've rewritten buffering code for `SystemTextJsonSerializer` code and added a minor optimization to avoid copying `NpgsqlStream` to `MemoryStream`.

oskardudycz reviewed (approved):
Wow, sweat work, @solyutor! Great contribution 👍 
Added one nitpicking comment. Could you have a look at it? 😉

solyutor commented on src/Marten/Events/Daemon/ActionParameters.cs:
```diff
@@ -54,15 +54,16 @@ public void IncrementAttempts(TimeSpan delay = default)
         }
 
-        public void ApplySkip(SkipEvent skip)
+        public async ValueTask ApplySkip(SkipEvent skip)
```
@oskardudycz,  I inspected all downstream calls: 4 of 5  run synchronously  except `ViewProjection.SliceAsyncEvents()`. The last has sync branches too.  According to these facts I believe it's better to keep `ValueTask` as a return value but rename to `ApplySkipAsync()`. 
It's possible to micro-optimize it as shown below but I doubt it worse 3 lines of code 

```csharp
        public async ValueTask ApplySkipAsync(SkipEvent skip)
        {
            if (Group != null)
            {
                ValueTask skipTask = Group.SkipEventSequence(skip.Event.Sequence);
                if (!skipTask.IsCompleted)
                {
                    await skipTask.ConfigureAwait(false);
                }

                // You have to reset the CancellationToken for the group
                Group.Reset();
                Cancellation = Group.Cancellation;
            }

            // Basically saying that the attempts start over when we skip
            Attempts = 0;
            Delay = default;
        }
```

oskardudycz replied:
Hmm, I might be missing something, but isn't it only used in Projection Daemon (precisely: https://github.com/JasperFx/marten/blob/master/src/Marten/Events/Daemon/ProjectionDaemon.cs#L402)? It's already an asynchronous method, so we should be good. Without that, all the sync methods would need to call `.Wait()` to be sure that there are no threading issues.

Thoughts?

solyutor replied:
Yes, `ActionParameters.SkipAsync()` is used in `ProjectionDaemon.TryAction()` only. Further calls may be async or sync (sync by nature, not need to `Task.Wait()` calls). Pseudo-graphics below shows possible call trees.
async Task ProjectionDaemon.TryAction() 
---> async ValueTask ActionParameters.ApplySkipAsync()
-----> abstract ValueTask [EventRangeGroup.SkipEventSequence()
------|=> async ValueTask [TenantSliceRange<TDoc, TId>.SkipEventSequence()](https://github.com/JasperFx/marten/blob/b5325056d6226159565036b0eda89d0b89ff5429/src/Marten/Events/Aggregation/TenantSliceRange.cs#L57)
---------> abstract ValueTask IEventSlicer.SliceAsyncEvents()
----------|=> sync ValueTask [ByStreamId.SliceAsyncEvents()](https://github.com/JasperFx/marten/blob/08efa3cfb31c3549f2449bb7da5cfd70a0bb2eb1/src/Marten/Events/Aggregation/ByStreamId.cs#L22)
----------|=> sync ValueTask [ByStreamKey<TDoc>.SliceAsyncEvents()](https://github.com/JasperFx/marten/blob/08efa3cfb31c3549f2449bb7da5cfd70a0bb2eb1/src/Marten/Events/Aggregation/ByStreamKey.cs#L20)
----------|=> async ValueTask [ViewProjection<TDoc, TId>.SliceAsyncEvents()](https://github.com/JasperFx/marten/blob/b5325056d6226159565036b0eda89d0b89ff5429/src/Marten/Events/Projections/ViewProjection.cs#L65)
------|=>--sync ValueTask [TenantedEventRange<TDoc, TId>.SkipEventSequence()](https://github.com/JasperFx/marten/blob/df12e5a065b82b9edd0497e9a08d2faedb7f5a33/src/Marten/Events/Daemon/TenantedEventRange.cs#L46)

I still think making signature `public async ValueTask ApplySkipAsync(SkipEvent skip)` is reasonable as it can be completed in a thread safe and synchronous way.

oskardudycz replied:
I'm not an async expert, but in my opinion, it's more on how do we intend to use it, not what methods we're using inside. 
At this stage, we already know that `ApplySkipAsync` is not used anywhere, and we do not intend to use it synchronously. Thus it'd be more straightforward to state that by the method's signature explicitly. When we have the need to use it synchronously (hopefully never 😅), we can change it to ValueTask.

It's not a blocker for me tho.

mysticmind replied:
`ApplySkipAsync` is always async in terms of usage and not going to be used synchronously. Hence agree with @oskardudycz that it is better for the return type to be `Task` which is more clear rather than `ValueTask`. That said, the class is `internal` so it is not a blocker by any means.

solyutor replied:
Ok, I have changed it to `Task`.

mysticmind reviewed (approved):
Looks good to me, great contribution @solyutor 👍 

Left a comment and was of the same opinion as @oskardudycz with regards to the return type for `ApplySkipAsync`.

I also ran the async daemon unit tests and all of them are green. See screenshot below:
<img width="612" alt="Screenshot 2021-11-06 at 7 46 33 PM" src="https://user-images.githubusercontent.com/2864208/140613364-45c43ab6-ff50-4fa0-b944-26021efcdcfa.png">
